### PR TITLE
L2 -> L1 messages in trace

### LINKED ISSUE #1752: Add L2 -> L1 messages to trace data
### Which component is your feature related to?

Forge

### Feature Request

.

### PR DESCRIPTION
**Pull request by war-in**
Closes #1752 

## Introduced changes

- l2 to l1 messages are saved to trace

## Checklist

- [x] Linked relevant issue
- [x] Updated relevant documentation
- [x] Added relevant tests
- [x] Performed self-review of the code
- [x] Added changes to `CHANGELOG.md`

### CODE REVIEW

File: crates/forge/tests/integration/trace.rs
Comment (piotmag769) [inline crates/forge/tests/integration/trace.rs:1]: Unnecessary changes

File: crates/cheatnet/src/state.rs [hunk 1]
```diff
@@ -328,14 +339,27 @@ impl TraceData {
         current_call.borrow_mut().entry_point.class_hash = Some(class_hash);
     }
 
-    pub fn exit_nested_call(&mut self, resources_used_after_call: &ExecutionResources) {
+    pub fn exit_nested_call(
+        &mut self,
+        resources_used_after_call: &ExecutionResources,
+        maybe_call_info: &Result<CallInfo, EntryPointExecutionError>,
```
Comment (piotmag769) [inline crates/cheatnet/src/state.rs:345]: If we can somehow reuse that in the PR adding the `CallResult` then it's fine

**Review by piotmag769 — APPROVED** [review]
Gj

File: crates/cheatnet/src/runtime_extensions/call_to_blockifier_runtime_extension/execution/entry_point.rs [hunk 2]
```diff
@@ -140,6 +141,17 @@ pub fn execute_call_entry_point(
...
+        result.as_ref().map_or_else(
+            |_| vec![],
+            |call_info| {
+                call_info
+                    .execution
+                    .l2_to_l1_messages
+                    .iter()
+                    .map(|ordered_message| ordered_message.message.payload.0.len())
+                    .collect()
+            },
```
Comment (piotmag769) [inline crates/cheatnet/src/runtime_extensions/call_to_blockifier_runtime_extension/execution/entry_point.rs:153]: Separate var pls 🙏

**Review by piotmag769 — APPROVED** [review]

File: crates/cheatnet/src/runtime_extensions/forge_runtime_extension/mod.rs [hunk 3]
```diff
@@ -765,6 +765,29 @@ pub fn update_top_call_execution_resources(runtime: &mut ForgeRuntime) {
     top_call.borrow_mut().used_execution_resources = all_execution_resources;
 }
 
+pub fn update_top_call_onchain_data(runtime: &mut ForgeRuntime) {
```
Comment (Arcticae) [inline crates/cheatnet/src/runtime_extensions/forge_runtime_extension/mod.rs:768]: This is done below in `get_all_used_resources` in a different manner
Why is this method needed?

File: crates/forge/tests/data/trace_resources/src/lib.cairo [hunk 4]
```diff
@@ -21,7 +21,8 @@ fn use_builtins_and_syscalls(empty_hash: ClassHash, salt: felt252) -> ContractAd
     get_block_hash_syscall(1).unwrap_syscall();
     starknet::deploy_syscall(empty_hash, salt, array![].span(), false).unwrap_syscall();
     emit_event_syscall(array![1].span(), array![2].span()).unwrap_syscall();
-    send_message_to_l1_syscall(10, array![20].span()).unwrap_syscall();
+    send_message_to_l1_syscall(10, array![20, 30].span()).unwrap_syscall();
+    send_message_to_l1_syscall(10, array![20, 30].span()).unwrap_syscall();
```
Comment (Arcticae) [inline crates/forge/tests/data/trace_resources/src/lib.cairo:25]: Why increase the number here? It influences all of the tests needlessly and makes the logic even more messy

**Review by Arcticae — CHANGES_REQUESTED** [review]

File: crates/cheatnet/src/runtime_extensions/forge_runtime_extension/mod.rs (hunk 3 above)
Comment (piotmag769) [reply crates/cheatnet/src/runtime_extensions/forge_runtime_extension/mod.rs:768]: It is not the same - in this function we add messages to the top call in the trace. In the one below we obtain payloads of all messages from all calls (mind the `inner_calls: starknet_runtime.hint_handler.inner_calls`) via blockifier method and add them to cumulative resources

File: crates/forge/tests/data/trace_resources/src/lib.cairo (hunk 4 above)
Comment (war-in) [reply crates/forge/tests/data/trace_resources/src/lib.cairo:25]: I just wanted to check if a few messages will be collected, but probably if the one is the rest will be too

File: crates/cheatnet/src/runtime_extensions/call_to_blockifier_runtime_extension/execution/entry_point.rs (hunk 2 above)
Comment (Arcticae) [reply crates/cheatnet/src/runtime_extensions/call_to_blockifier_runtime_extension/execution/entry_point.rs:153]: can we pass `call_info.execution.l2_to_l1_messages` instead and process it down the line? it makes code less readable here

**Comment by war-in** [comment]
https://github.com/software-mansion/cairo-profiler/pull/35

wait for this to be merged

**Review by piotmag769 — APPROVED** [review]
LGTM, but we prob want to wait for @Arcticae to merge the blockifier update first (higher prio), so he doesn't have to resolve more conflicts

**Comment by piotmag769** [comment]
@Arcticae then it needs your approve I guess xD

**Review by Arcticae — APPROVED** [review]
